Python/Product/AnacondaInstallLauncher/Program.cs

```diff
@@ -0,0 +1,212 @@
+            var psi = new ProcessStartInfo(filename, arguments) {
+                CreateNoWindow = true,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false
+            };
+            Console.Error.WriteLine("Starting:{0}  Filename: {1}{0}  Argument: {2}", Environment.NewLine, psi.FileName, psi.Arguments);
+            var p = Process.Start(psi);
+
+            p.OutputDataReceived += (_, e) => { if (e.Data != null) Console.Write(e.Data); };
+            p.ErrorDataReceived += (_, e) => { if (e.Data != null) Console.Error.Write(e.Data); };
+
+            return p;
+        }
+
+        int DoInstall() {
+            WaitForUninstall(false);
+            if (_cancel) {
+                return ERROR_INSTALL_CANCEL;
+            }
+
+            var p = Run(_installer, string.Format(CultureInfo.InvariantCulture, "/InstallationType=AllUsers /RegisterPython=0 /S /D={0}", _targetDir));
+            p.WaitForExit();
+            return p.ExitCode;
+        }
+
+        int DoUninstall() {
+            var p = Run(_installer, "/S");
+
+            if (p.WaitForExit(30000)) {
+                Console.Error.WriteLine("Process exited: {0}", p.ExitCode);
+            } else {
+                Console.Error.WriteLine("Process has not exited");
+            }
+
+            WaitForUninstall(true);
+
+            return _cancel ? ERROR_INSTALL_CANCEL : p.ExitCode;
+        }
+
+        void WaitForUninstall(bool requireWait) {
+            try {
+                WaitForProcess(DateTime.UtcNow.AddMinutes(30), requireWait);
+            } catch (Exception ex) {
+                Console.Error.WriteLine(ex.ToString());
+            }
+        }
+
+        bool KeepWaiting {
+            get {
+                if (_cancel) {
+                    return false;
+                }
+
+                if (!Directory.Exists(_targetDir)) {
+                    Console.Error.WriteLine("Target directory is not present");
+                    return false;
+                }
+                try {
+                    if (!Directory.EnumerateFileSystemEntries(_targetDir).Any()) {
+                        Console.Error.WriteLine("Target directory is empty");
+                        return false;
+                    }
+                } catch (Exception ex) {
+                    Console.Error.WriteLine("Error reading target directory");
+                    Console.Error.WriteLine(ex.ToString());
+                    return false;
+                }
+
+                return true;
+            }
+        }
+
+        Process[] GetProcesses() {
+            return Process.GetProcesses()
+                .Where(p => p.ProcessName.StartsWith("Un_") &&
+                    p.MainModule.FileVersionInfo.FileDescription.IndexOf("Anaconda", StringComparison.OrdinalIgnoreCase) >= 0)
+                .ToArray();
+        }
+
+        void WaitForProcess(DateTime stopAt, bool requireProcess) {
```

Comment:
Why timeout? If you use the process handle (via `Process.WaitForExit`), this is an intrinsic wait handle (in the underlying native implementation). It will signal when the process terminates. If you merely timeout, the original problem of the race condition between the async uninstaller and sync installer will happen as well.

If you really feel it necessary to timeout (and with how much machines vary in performance, especially under additional workload, I don't recommend it), at least fail method and forego running the installer. Return an exit code so other chainers (like Visual Studio) can tell the user something failed and to repair.

Reply:
Yeah, fair enough. I'll lose the timeout.